### PR TITLE
Migrated root build.gradle to Kotlin

### PR DESCRIPTION
**Pull request by hansd**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    allModules = configurations["externalModulesRuntime"]
+    coreModules = configurations["coreRuntime"]
+    modulesToPatch = configurations["externalModules"]
+    destination = patchedExternalModulesDir
+}
+
+evaluationDependsOn(":distributions")
+
+tasks.create<Install>("install") {
+    description = "Installs the minimal distribution into directory ${project.property("gradle_installPath")}"
```
Comment (blindpirate) [inline build.gradle.kts]: `${project.property("gradle_installPath")}` will fail the configuration phase if `gradle_installPath` property not set.

Comment (eskatos) [inline build.gradle.kts]: :x: `$gradle_installPath` should not be evaluated in this description
👍 to @blindpirate's [comment](https://github.com/gradle/gradle/pull/4851/files?diff=unified#r178696438).
Note that to escape `$` in a Kotlin string you can use `${"$"}`

Also applies to the `installAll` task below

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+extra.set("allTestRuntimeDependencies", configurations["testRuntime"].allDependencies)
+
+val patchedExternalModulesDir = File(buildDir, "external/files")
+val patchedExternalModules = files(Callable { fileTree(patchedExternalModulesDir).files.sorted() })
+patchedExternalModules.builtBy("patchExternalModules")
+
+dependencies {
+    "externalModules"("org.gradle:gradle-kotlin-dsl:${BuildEnvironment.gradleKotlinDslVersion}")
```
Comment (eskatos) [inline build.gradle.kts]: :o: consider bringing all the `Configuration`s into scope when creating them so there's no need to repeat strings to reference them here.

e.g.
```kotlin
val externalModules by configurations.creating {
    visible = false
}
// ...
dependencies {
    externalModules("...")
}
```

File: build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+    all {
+        attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage::class.java, Usage.JAVA_RUNTIME))
+    }
+}
+
+extra.set("allTestRuntimeDependencies", configurations["testRuntime"].allDependencies)
+
+val patchedExternalModulesDir = File(buildDir, "external/files")
+val patchedExternalModules = files(Callable { fileTree(patchedExternalModulesDir).files.sorted() })
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply build.gradle.kts]: This API is something I wish I'd known about.

File: build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+tasks.create<Install>("installAll") {
+    description = "Installs the full distribution into directory ${project.property("gradle_installPath")}"
+    group = "build"
+    with(project(":distributions").property("allDistImage") as CopySpec)
+    installDirPropertyName = "gradle_installPath"
+}
+
+afterEvaluate {
+    if (gradle.startParameter.isBuildCacheEnabled()) {
+        rootProject.availableJavaInstallations.validateBuildCacheConfiguration((gradle as GradleInternal).settings.buildCache)
```
Comment (JLLeitschuh) [inline build.gradle.kts]: :o: Seems weird that your own Gradle build has to use an internal API to achieve the logic you want.

File: build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    "gradlePlugins" {
+        isVisible = false
+    }
+    "testRuntime" {
+        extendsFrom(configurations["runtime"])
+        extendsFrom(configurations["gradlePlugins"])
+    }
+
+    all {
+        attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage::class.java, Usage.JAVA_RUNTIME))
```
Comment (bamboo) [inline build.gradle.kts]: :o: Doesn't this `attribute` declaration make previous `attribute` declarations unnecessary?

File: build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+afterEvaluate {
+    if (gradle.startParameter.isBuildCacheEnabled) {
+        rootProject
+            .availableJavaInstallations
+            .validateBuildCacheConfiguration(buildCacheConfiguration())
+    }
+}
+
+fun Project.buildCacheConfiguration() =
+    (gradle as GradleInternal).settings.buildCache
```
Comment (JLLeitschuh) [inline build.gradle.kts]: :o: Same comment as before.
Seems weird you are using an internal API in your own build.

**Comment by bamboo** [comment]
[CI build](https://builds.gradle.org/viewQueued.html?itemId=12127479&tab=queuedBuildOverviewTab)

**Pull request merged**